### PR TITLE
Support all query types in DNS mode

### PR DESCRIPTION
**Pull request by errorxyz**
#### Description

Uses `mitmproxy_rs.DnsResolver` to resolve domain names that need to be checked in the hosts file first and forwards any other type of query to the server specified in the options
Fixes https://github.com/mitmproxy/mitmproxy/issues/6739

#### Checklist

 - [x] I have updated tests where applicable.
 - [x] I have added an entry to the CHANGELOG.

~TODO: check and fix tests after we release latest version of `mitmproxy_rs`~

### CODE REVIEW

**Comment by errorxyz** [comment]
How do I tell the tests to use the updated `mitmproxy_rs`?

**Comment by mhils** [comment]
> How do I tell the tests to use the updated `mitmproxy_rs`?

This currently requires us to ship a new release of mitmproxy_rs. There is an upper bound in https://github.com/mitmproxy/mitmproxy/blob/main/pyproject.toml#L45 though, so we can safely do that without affecting our users.

**Comment by mhils** [comment]
Regarding this PR: Could we just expose a `get_system_dns_server()` from Rust (that uses https://docs.rs/hickory-resolver/latest/hickory_resolver/system_conf/fn.read_system_conf.html internally) instead? We can then treat DNS mode just like we treat reverse DNS mode.

(We still need our getaddrinfo for mitmproxy.proxy.server - but the addon here we can maybe simplify dramatically 😃)

**Comment by errorxyz** [comment]
> that uses https://docs.rs/hickory-resolver/latest/hickory_resolver/system_conf/fn.read_system_conf.html internally

This is available for UNIX only

**Comment by mhils** [comment]
> > that uses https://docs.rs/hickory-resolver/latest/hickory_resolver/system_conf/fn.read_system_conf.html internally
> 
> This is available for UNIX only

The docs are misleading here - it's available for Windows as well if you look at the sources. :)

File: mitmproxy/addons/dns_resolver.py [hunk 1]
```diff
@@ -155,7 +38,85 @@ async def dns_request(self, flow: dns.DNSFlow) -> None:
...
+            "use_hosts_file",
+            bool,
+            True,
+            "Use the hosts file for DNS lookups in regular DNS mode/wireguard mode.",
+        )
+
+        loader.add_option(
+            "name_servers",
+            Sequence[str],
+            mitmproxy_rs.get_system_dns_servers(),
```
Comment (mhils) [inline mitmproxy/addons/dns_resolver.py:108]: Let's default this to empty list and then do `ctx.options.name_servers or mitmproxy_rs.get_system_dns_servers()` below. Two reasons:

1. get_system_dns_servers() may potentially raise.
2. The config in `_options_gen.ts` shouldn't be platform specific.

**Review by mhils — COMMENTED** [review]
Looks great, thanks! I'm curious whether we will run into any issues with overly long records or so, but for now this is a great way forward. 👍

File: mitmproxy/addons/dns_resolver.py (hunk 1 above)
Comment (errorxyz) [reply mitmproxy/addons/dns_resolver.py:108]: Do you think we should implement a `Sequence[str] | None` option type for this case?

**Comment by errorxyz** [comment]
> I'm curious whether we will run into any issues with overly long records or so, but for now this is a great way forward

I don't think we'll run into this issue practically since we're only handling A/AAAA records(they typically won't cross 512 bytes). That TODO is a remnant of the old code and theoretically speaking, we *may* have a response size greater than 512 bytes that would need to be handled.

File: mitmproxy/addons/dns_resolver.py [hunk 2]
```diff
@@ -155,7 +67,56 @@ async def dns_request(self, flow: dns.DNSFlow) -> None:
...
+        try:
+            if question.type == dns.types.A:
+                addrinfos = await self.resolver.lookup_ipv4(question.name)
+            elif question.type == dns.types.AAAA:
+                addrinfos = await self.resolver.lookup_ipv6(question.name)
+        except socket.gaierror as e:
+            if e.args[0] == "NXDOMAIN":
+                raise ResolveError(dns.response_codes.NXDOMAIN)
+            else:  # pragma: no cover
+                raise ResolveError(dns.response_codes.SERVFAIL)
```
Comment (errorxyz) [inline mitmproxy/addons/dns_resolver.py:111]: How do we want to handle `NOERROR` codes here i.e `if e.args[0] == "NOERROR"`? 
(we can get a `NOERROR` here if there are no records returned but the name is valid)
https://datatracker.ietf.org/doc/html/rfc2308#section-2.2

File: mitmproxy/addons/dns_resolver.py [hunk 3]
```diff
@@ -19,129 +19,41 @@ def __init__(self, response_code: int) -> None:
...
-        dns.ResourceRecord(
-            name=question.name,
-            type=question.type,
-            class_=question.class_,
-            ttl=dns.ResourceRecord.DEFAULT_TTL,
-            data=dns.domain_names.pack(name),
+class DnsResolver:
+    def load(self, loader):
+        loader.add_option(
+            "use_hosts_file",
```
Comment (mhils) [inline mitmproxy/addons/dns_resolver.py:25]: We generally suck at this, but maybe we should prefix the new options with `dns_`?

File: mitmproxy/addons/dns_resolver.py (hunk 2 above)
Comment (mhils) [reply mitmproxy/addons/dns_resolver.py:111]: Then we should also send a NOERROR to the client, no?

File: test/mitmproxy/addons/test_dns_resolver.py [hunk 4]
```diff
@@ -32,107 +25,86 @@ async def test_simple(monkeypatch):
...
-        lambda rr: rr.domain_name == "dns.google",
+        res_old = dr.resolver
+        tctx.options.use_hosts_file = False
+        assert dr.resolver != res_old
+
+
+async def lookup_ipv4(name: str):
+    if name == "not.exists":
+        raise socket.gaierror("NXDOMAIN")
+    return await asyncio.sleep(0, ["8.8.8.8"])
```
Comment (mhils) [inline test/mitmproxy/addons/test_dns_resolver.py:44]: Not sure if I want to know - what is this sleep here about? 😄

File: test/mitmproxy/addons/test_dns_resolver.py [hunk 5]
```diff
@@ -1,25 +1,18 @@
...
 from mitmproxy.test import tflow
 from mitmproxy.test import tutils
 
 
-async def test_simple(monkeypatch):
-    monkeypatch.setattr(
-        dns_resolver, "resolve_message", lambda _, __: asyncio.sleep(0, "resp")
-    )
-
+async def test_simple():
```
Comment (mhils) [inline test/mitmproxy/addons/test_dns_resolver.py:15]:
```suggestion
async def test_ignores_reverse_mode():
```

**Review by mhils — COMMENTED** [review]
LGTM % nits! 😃

File: mitmproxy/addons/dns_resolver.py (hunk 2 above)
Comment (errorxyz) [reply mitmproxy/addons/dns_resolver.py:111]: What about the authority section? Can we ignore that?

File: test/mitmproxy/addons/test_dns_resolver.py (hunk 4 above)
Comment (errorxyz) [reply test/mitmproxy/addons/test_dns_resolver.py:44]: oops, my bad

File: mitmproxy/addons/dns_resolver.py [hunk 6]
```diff
@@ -19,129 +19,41 @@ def __init__(self, response_code: int) -> None:
...
+    def configure(self, updated):
+        if "use_hosts_file" in updated or "name_servers" in updated:
+            del self.resolver
+            del self.name_servers
+
+    @cached_property
+    def resolver(self) -> mitmproxy_rs.DnsResolver:
+        return mitmproxy_rs.DnsResolver(
+            name_servers=ctx.options.name_servers
+            or mitmproxy_rs.get_system_dns_servers(),
```
Comment (errorxyz) [inline mitmproxy/addons/dns_resolver.py:47]:
```suggestion
            name_servers=self.name_servers
```
Adding this change along with the suggested ones

File: mitmproxy/addons/dns_resolver.py [hunk 7]
```diff
@@ -19,129 +19,41 @@ def __init__(self, response_code: int) -> None:
...
+            "Name servers to use for lookups. Default: operating system's name servers",
         )
-    elif question.type == dns.types.AAAA:
-        return await resolve_question_by_name(
-            question, loop, socket.AddressFamily.AF_INET6, ipaddress.IPv6Address
+
+    def configure(self, updated):
+        if "use_hosts_file" in updated or "name_servers" in updated:
+            del self.resolver
+            del self.name_servers
```
Comment (errorxyz) [inline mitmproxy/addons/dns_resolver.py:41]: Changing both options at the same time results in an `AttributeError` perhaps we want to use:
```suggestion
    def configure(self, updated):
        if "use_hosts_file" in updated or "name_servers" in updated:
            self.__dict__.pop('resolver', None)
            self.__dict__.pop('name_servers', None)
```

File: mitmproxy/addons/dns_resolver.py (hunk 2 above)
Comment (mhils) [reply mitmproxy/addons/dns_resolver.py:111]: Yes, I'd say let's just ignore that for now. :)

**Comment by mhils** [comment]
https://github.com/mitmproxy/mitmproxy/pull/6975#discussion_r1676522249, otherwise LGTM :)

**Review by mhils — APPROVED** [review]

**Comment by errorxyz** [comment]
Sorry, messed up the commit history a bit while hopping between windows and linux. Should be fixed now

**Comment by mhils** [comment]
No worries, this is why we squash! :)

**Pull request merged**